src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningConditionMapper.java

```diff
@@ -0,0 +1,367 @@
+
+        private final PrimaryKey primaryKey;
+        private final Map<String, String> fieldPlaceholders;
+        private final Map<String, AttributeValue> valuePlaceholders;
+
+        // output
+        private AttributeValue hkValue = null;
+        private Optional<KeyFieldCondition> rkCondition = Optional.empty();
+
+        KeyConditionExpressionVisitor(PrimaryKey primaryKey, Map<String, String> fieldPlaceholders,
+                                      Map<String, AttributeValue> valuePlaceholders) {
+            this.primaryKey = primaryKey;
+            this.fieldPlaceholders = fieldPlaceholders;
+            this.valuePlaceholders = valuePlaceholders;
+        }
+
+        @Override
+        public Void visitKeyConditionPart(KeyConditionPartContext keyConditionPart) {
+            IdContext id = keyConditionPart.id();
+            if (id != null) {
+                String fieldName = keyConditionPart.id().getText();
+                if (fieldName.startsWith("#")) {
+                    // if it's a field placeholder, remove it -- we'll generated a new one for translated expression
+                    fieldName = removeFieldPlaceholder(fieldName);
+                }
+
+                if (primaryKey.getHashKey().equals(fieldName)) {
+                    checkArgument(hkValue == null, "Multiple hash key conditions");
+                    ComparatorContext comparator = keyConditionPart.comparator();
+                    checkArgument(comparator != null && comparator.getText().equals("="),
+                        "Hash key condition must have operator '='");
+                    String valuePlaceholder = keyConditionPart.literal(0).getText();
+                    hkValue = removeValuePlaceholder(valuePlaceholder);
+                } else if (primaryKey.getRangeKey().isPresent() && primaryKey.getRangeKey().get().equals(fieldName)) {
+                    checkArgument(rkCondition.isEmpty(), "Multiple range key conditions");
+                    ComparatorContext comparator = keyConditionPart.comparator();
+                    ComparisonOperator op = comparator != null ? getComparatorType(comparator.getText()) : BETWEEN;
+                    List<AttributeValue> values = keyConditionPart.literal().stream()
+                        .map(LiteralContext::getText)
+                        .map(this::removeValuePlaceholder)
+                        .collect(Collectors.toList());
+                    rkCondition = Optional.of(new KeyFieldCondition(op, values));
+                } else {
+                    throw new IllegalArgumentException("Key condition expression contains non-key field: " + fieldName);
+                }
+            }
+            return visitChildren(keyConditionPart);
+        }
+
+        private AttributeValue removeValuePlaceholder(String placeholder) {
+            AttributeValue value = null;
+            if (valuePlaceholders != null) {
+                value = valuePlaceholders.remove(placeholder);
```

Comment:
What happens if the map passed in by the client of mt-dynamo contains placeholders that do not appear in expressions?

Reply:
Oh good point... yeah I think this is a bug. This value can also appear in the query's filter expression, in which case we can't remove it...

Or are you asking what the underlying Dynamo would do if there is a field or value placeholder that doesn't appear in any expressions in the request? At least according to behavior I've seen from DynamoDB Local, the request will be rejected in that case.